servicetalk-encoding-api/src/main/java/io/servicetalk/encoding/api/AbstractZipContentCodec.java

```diff
@@ -579,4 +580,55 @@ public void write(byte[] b, int off, int len) {
             buffer.writeBytes(b, off, len);
         }
     }
+
+    static final class BufferBoundedInputStream extends InputStream {
+        private final Buffer buffer;
+        private final int limit;
+        private int count;
+
+        BufferBoundedInputStream(Buffer buffer, int limit) {
+            this.buffer = requireNonNull(buffer);
+            this.limit = limit;
+        }
+
+        @Override
+        public int read() {
+            if (buffer.readableBytes() == 0) {
+                return -1;
+            }
+            if (++count > limit) {
+                throw new IndexOutOfBoundsException("Buffer limit has been reached: " +
+                        count + " (expected <= " + limit + ") bytes");
+            }
+            return buffer.readByte() & 0xff;
+        }
+
+        @Override
+        public int read(byte[] b, int off, int len) {
+            int readableBytes = buffer.readableBytes();
+            if (readableBytes == 0) {
+                return -1;
+            }
+            int bytes = min(readableBytes, len);
+            if (count + bytes > limit) {
+                throw new IndexOutOfBoundsException("Buffer limit has been reached: " +
+                        count + " (expected <= " + limit + ") bytes");
+            }
+            count += bytes;
+            buffer.readBytes(b, off, bytes);
+            return bytes;
+        }
+
+        @Override
+        public long skip(long n) {
+            int skipped = min(buffer.readableBytes(), (int) min(Integer.MAX_VALUE, n));
```

Comment:
I added the negative guard, but the cast to integer might not be the right way.
Currently the max allowed input is `INT_MAX`, if I relax that and cast the input, that could mean that an arbitrary input of (`long 4294967297` -> `int 1`) can be used to skip 1 bytes rather than all available bytes. 
Unexpected behavior.

Reply:
ofc 🤦